### PR TITLE
DP-8339: Automate Mayflower Release Process

### PR DESCRIPTION
**Pull request by rbayliss**
Any PRs being created needs a changelog.txt file before being merged into dev. See: [Change Log Instructions](https://github.com/massgov/mayflower/blob/dev/docs/change-log-instructions.md)

## Description

The high level goal of this ticket is to make Mayflower releases much simpler.  We should be able to trigger all of the release steps (NPM, S3, Artifacts) just by pushing a tag to this repository, and letting CircleCI do the work.  This PR implements this automation.

Successful Build: https://circleci.com/workflow-run/d3b50893-cee6-4eb1-94e0-6cce5dcfd7a7
This build released the 5.17.0-alpha1 tag to the mayflower-artifacts repo and NPM (see https://www.npmjs.com/package/@massds/mayflower).  It didn't update the root S3 installation because it's not a stable tag (only stable tags are released to S3's / directory).

 Note that Visual Regression testing shows **0** differences here, so we're really only concerned with how this integrates with other applications.  There ended up being no template or CSS changes, and only one minor JS change required, so as long as it works with Drupal and Mayflower React, this should be a no-impact change, despite the intimidating list below. 

### Changes:
* Updates to all build-only dependencies (didn't touch neat, bourbon, or babel) including Gulp 4.
* Refactor gulp tasks on Gulp 4 and split into 3 distinct "task registries".
* Release built Pattern Lab display for stable tags to /v/${MAJOR_VERSION}
* Release built Pattern Lab display for branches to /b/${BRANCH}
* Exclude moment JS from custom JS bundle (saves 100k!).
* Additional minification for vendor JS bundle.
* Cleans mayflower-artifacts so it actually removes files that are removed on the source side.  This removes a lot of [old leftovers](https://github.com/massgov/mayflower-artifacts/compare/DP-8339-mayflower_release_revamp)
* Move to babel-preset-env, which is configured in package.json.
* Use a .browserslistrc for determining the list of supported browsers for autoprefixer.  We can eventually use this for babel, too, but need to update some of our code to work with newer versions of babel.
* Change most asset imports to be relative (should only affect the _meta templates).
* Determine the URL from an environment variable in Pattern Lab.
* Determine the Release from package.json in Pattern Lab.
* Determine the Release date from the date of the last commit in Pattern Lab.

#### Pattern Lab
* `patternlab:build` - Compile Pattern Lab into `styleguide/public`
* `patternlab:watch` - Compile and start a live development server.
* `patternlab:publish` - Compile and deploy to the S3 bucket at /b/${BRANCH}.
* `patternlab:release` - Compile and deploy to the S3 bucket at /b/${BRANCH}.  If tagged, and tag is not a pre-release tag, also deploys to /v/${MAJOR_VERSION}, and /.

#### Artifacts
* `artifacts:build` - Clones artifact repo, cleans it, and builds assets into `styleguide/artifacts`
* `artifacts:publish` - Runs `artifacts:build`, then pushes to `$BRANCH` on the artifacts repository.
* `artifacts:release` - Runs `artifacts:build`, then pushes to `$BRANCH` and `$TAG` on the artifacts repository.

#### NPM
* `npm:build` - Builds the NPM package to `styleguide/npm`.
* `npm:release` - Runs `npm:build` and then publishes to the NPM registry.

## Open Questions:
* Do we still use the `svg2twig` and `svg-sprite` Gulp tasks?

### Followups needed:
* Remove old `/scripts` and Phing build after this work has been proven reliable.
* Automated cleanup of S3 bucket.
* Update Mayflower to latest babel and use `babel-preset-env` with `.browserslistrc`
* Review README.md and developer-deploy.md and make sure they still make sense.

### Once this is merged:
* Announce to #massgov:
    > Attention Mayflower developers: we have updated the way Mayflower is generated locally and in CircleCI.  Please be aware of the following changes.
    * You should run `npm install` before beginning work or you will get errors running Gulp.
    * Pre-release tags are no longer necessary for previewing changes in Drupal.  You can now preview a branch instead of needing a tag.  To do this, run the following command from the Drupal repository after pushing your changes to a Mayflower branch: `composer require palantirnet/mayflower:dev-DP-1000-my-branch`, where "DP-1000-my-branch" is the name of your branch.
    * The Mayflower release process is now handled mostly by CircleCI.  If you need to do a Mayflower release, please read the updated release instructions (https://github.com/massgov/mayflower/blob/dev/docs/release.md).

## Related Issue / Ticket

- [JIRA issue](https://jira.state.ma.us/browse/DP-8339)

## Steps to Test

1. Code review is probably the most important here.
2. Try setting up for local development and make sure you can still do all the same things.
2. Check the [deployed Pattern Lab instance](https://mayflower.digital.mass.gov/b/DP-8339-mayflower_release_revamp/index.html)
3. Check the [artifacts release tag](https://github.com/massgov/mayflower-artifacts/releases/tag/5.17.0-alpha1).  This is what will be pulled into Drupal.
4. Check the [NPM package](https://www.npmjs.com/package/@massds/mayflower). This is what will be pulled into Mayflower React.

## Screenshots
Use something like [licecap](http://www.cockos.com/licecap/) to capture gifs to demonstrate behaviors.

## Additional Notes:

Anything else to add?

#### Impacted Areas in Application

* 

#### @TODO

*

### CODE REVIEW

**Review by weitzman — APPROVED** [review]
Excellent work!

- The [Rollback section](https://github.com/massgov/mayflower/blob/c741fdb991593f5de0f86507d883731e61c6d92b/docs/release.md#rollback) is too simple? There are no changes to the artifacts repo in a rollback? 
 
A couple more followups:
- Review README.md for updates
- Review docs/developer-deploy.md for updates

**Comment by rbayliss** [comment]
@weitzman - I went with the "minimal changes" approach on the rollback documentation, but I agree that the instructions are inadequate.  Honestly, I'm not sure we should even have a rollback process.  Once a release is out, it's out.  Unpublishing it from the artifacts and NPM is probably not an option, and could cause more breakage than it fixes (if someone has already pulled in the release).  With Composer and NPM, the responsibility for rollback is on the implementors (although it's Mayflower's responsibility to fix quickly).  If anyone feels strongly that we should have a rollback process beyond this, I'm happy to open a followup. 

Adding followup tasks for the README and developer-deploy docs.

**Comment by weitzman** [comment]
Yeah, I think there shouldnt be a rollback process. You have to revert
commits and make a new release.

**Comment by gender404** [comment]
I will second the new release over a rollback.  Once you've created a release tag (not a pre-release) that code could be consumed before a rollback takes place.  This happened to me with GitKraken and they rolled the code back and re-released to the original tag.  My software updated to the broken code release tag and then wouldn't pull the new code, because it already had the most recent tag.

You should push another release and tag instead of rolling back.

**Review by clairesunstudio — APPROVED** [review]
Work as described, tested pre-release, artifacts and the node package, looks great!

File: .circleci/config.yml [hunk 1]
```diff
@@ -64,37 +62,41 @@ jobs:
...
+            branches: { ignore: /.*/ }
+            tags: { only: /.*/ }
+      - deploy_tag:
           requires: [build]
           filters:
-            branches:
-              ignore: /^master$/
\ No newline at end of file
+            branches: { ignore: /.*/ }
+            tags: { only: /.*/ }
```
Comment (tnathanjames) [inline .circleci/config.yml]: A file-ending \n might make git happy.

File: docs/release.md [hunk 2]
```diff
@@ -7,13 +7,12 @@ Mayflower release managers with the necessary repo permissions can follow these
...
-1. Bump the version on the homepage - `@pages/readme2.json` - by updating the version and date text in `errorPage.type`
-1. Bump the version of the npm package by running `gulp bump -v=#.#.#` where `#.#.#` is the version you are releasing.
-1. Commit your version bump file updates.
+1. Move into the `styleguide` directory `cd styleguide` and run `npm install` in case the release includes new packages.
+1. Update the version of the npm package by editing the `version` field in `package.json`.
+1. Commit your version bump file update.
 1. Push release branch to `massgov/mayflower` (i.e. `git push upstream release-#.#.#`).
 1. Wait for [the circle build](https://circleci.com/gh/massgov/mayflower) to pass, which will deploy your release branch to staging automagically :).
-1. Verify release notes against the site rendered at: `https://mayflower.digital.mass.gov/<your-release-branch>/index.html`.
+1. Verify release notes against the site rendered at: `https://mayflower.digital.mass.gov/b/<your-release-branch>/index.html`.
```
Comment (tnathanjames) [inline docs/release.md:15]: /b/ is intentional?

Comment (tnathanjames) [reply docs/release.md:15]: I see below how you are using /b/.  Never mind here.

File: styleguide/.browserslistrc [hunk 3]
```diff
@@ -0,0 +1,7 @@
+# Browsers we support:
+
+> 1%
+last 3 versions
+Firefox ESR
+Opera 12.1
+ie >= 9
```
Comment (tnathanjames) [inline styleguide/.browserslistrc]: End of file \n would make git happy

File: styleguide/config/listeners.json [hunk 4]
```diff
@@ -1 +1 @@
-{ "listeners": [ ] }
\ No newline at end of file
+{ "listeners": ["\\PatternLab\\DynamicDataListener" ] }
```
Comment (tnathanjames) [inline styleguide/config/listeners.json]: End of file \n would make git happy

File: styleguide/core/src/PatternLab/DynamicDataListener.php [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+  private function getGitDate($ref) {
+    $proc = new Process('git show -s --format=%ci ' . escapeshellarg($ref));
+    $proc->run();
+    if($proc->isSuccessful()) {
+      $date = new \DateTime($proc->getOutput());
+      return $date->format('n/j/Y');
+    }
+    return FALSE;
+  }
+}
```
Comment (tnathanjames) [inline styleguide/core/src/PatternLab/DynamicDataListener.php]: End of file \n would make git happy

File: styleguide/source/assets/js/helpers/listing.js [hunk 6]
```diff
@@ -733,4 +732,4 @@ export default  function(window, document, undefined, $){
     makeMoment
   };
 
-}(window, document, undefined, jQuery);
+}(window, document, undefined, jQuery, moment);
```
Comment (tnathanjames) [inline styleguide/source/assets/js/helpers/listing.js:735]: It looks like the makeMoment method above uses moment.  Should makeMoment be removed as well?

Comment (rbayliss) [reply styleguide/source/assets/js/helpers/listing.js:735]: We're still using moment in this file, we're just not directly importing it anymore.  It's being injected from the global moment variable, which is made available in `vendors-generated.js`

File: styleguide/tools/gulp/Artifacts.js [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    }
+    resolveDest(subPath) {
+        if(!this.config.dest.artifact) {
+            throw new Error("Please set config.dest.artifact");
+        }
+        return subPath ? path.resolve(this.config.dest.artifact, subPath) : this.config.dest.artifact;
+    }
+}
+
+module.exports = ArtifactRegistry;
```
Comment (tnathanjames) [inline styleguide/tools/gulp/Artifacts.js]: An end of file line break would make git happy.

File: styleguide/tools/gulp/Base.js [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+            .pipe(js.custom(this.config.minify))
+            .pipe(gulp.dest(dest))
+            .pipe(this.debug(name));
+        task.displayName = name;
+        return task;
+    }
+
+}
+
+module.exports = MayflowerTaskRegistry;
```
Comment (tnathanjames) [inline styleguide/tools/gulp/Base.js]: An end of file line break would make git happy.

File: styleguide/tools/gulp/NPM.js [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+    }
+    resolveDest(subPath) {
+        if(!this.config.dest.npm) {
+            throw new Error("Please set config.dest.npm");
+        }
+        return subPath ? path.resolve(this.config.dest.npm, subPath) : this.config.dest.npm;
+    }
+}
+
+module.exports = NPMRegistry;
```
Comment (tnathanjames) [inline styleguide/tools/gulp/NPM.js]: An end of file line break would make git happy.

File: styleguide/tools/gulp/PatternLab.js [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+    }
+    resolveDest(subPath) {
+        if(!this.config.dest.patternlab) {
+            throw new Error("Please set config.dest.patternlab");
+        }
+        return subPath ? path.resolve(this.config.dest.patternlab, subPath) : this.config.dest.patternlab;
+    }
+}
+
+module.exports = PatternLabRegistry;
```
Comment (tnathanjames) [inline styleguide/tools/gulp/PatternLab.js]: An end of file line break would make git happy.

File: styleguide/tools/gulp/gulpfile.js [hunk 11]
```diff
@@ -1,128 +1,65 @@
...
-    dest: path.resolve(__dirname, "../../public/assets"),
-    patternLabRoot: path.resolve(__dirname, "../../"),
-    rootSite: path.resolve(__dirname, "../../public"),
-    rootSource: path.resolve(__dirname, "../../source"),
-    tasks: ["copy", "copyTwig", "js", "css", "bower", "patternlab"],
-    env: "development", // "development", "production", "local"
-    watch: false,
-    browserSync: false,
-    version: "0.0.0"
+var argv = require("optimist").argv;
```
Comment (tnathanjames) [inline styleguide/tools/gulp/gulpfile.js]: I believe this could be a const.

File: styleguide/tools/gulp/gulpfile.js [hunk 12]
```diff
@@ -1,128 +1,65 @@
...
+// @todo: Do we need svg2twig?
+// @todo: Do we need the svg-sprite task?
 
-  quench.build(config);
-});
 
-// watch for single tasks on the command line, eg "gulp js"
-quench.singleTasks(defaults);
+gulp.task("default", gulp.series("patternlab:watch"));
+gulp.task("prod", gulp.series("patternlab:build"));
```
Comment (tnathanjames) [inline styleguide/tools/gulp/gulpfile.js]: An end of file line break would make git happy.

File: styleguide/tools/gulp/helpers/escape.js [hunk 13]
```diff
@@ -0,0 +1,4 @@
+
+// Shell escape function found on stack overflow:
+// https://stackoverflow.com/a/22827128
+module.exports = (s) => "'" + s.replace(/'/g, "'\\''") + "'";
```
Comment (tnathanjames) [inline styleguide/tools/gulp/helpers/escape.js]: An end of file line break would make git happy.

File: styleguide/tools/gulp/helpers/git.js [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+    },
+    currentMessage: function() {
+        try {
+            return exec("git log -1 --pretty=%B 2> /dev/null").toString().trim();
+        }
+        catch(err) {
+            return;
+        }
+    }
+};
```
Comment (tnathanjames) [inline styleguide/tools/gulp/helpers/git.js]: An end of file line break would make git happy.

File: styleguide/tools/gulp/pipelines/js.js [hunk 15]
```diff
@@ -0,0 +1,75 @@
+var uglify         = require("gulp-uglify"),
```
Comment (tnathanjames) [inline styleguide/tools/gulp/pipelines/js.js]: Could be const.

File: styleguide/tools/gulp/pipelines/js.js [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+            .pipe(sourcemaps.init, {loadMaps: true})
+            .pipe(concat, "vendor.js", {newLine: ";"})
+            .pipe(function() {
+                return gulpIf(minify, uglify());
+            })
+            .pipe(rename, {suffix: "-generated"})
+            .pipe(sourcemaps.write, "./")();
+    },
+    custom: function(minify) {
+        var browserifyOptions = {
```
Comment (tnathanjames) [inline styleguide/tools/gulp/pipelines/js.js]: Could be const.

File: styleguide/tools/gulp/pipelines/js.js [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+            .pipe(sourcemaps.write, "./")();
+    }
+
+};
+
+function browserifyNoExternals(options) {
+    return through.obj(function (file, enc, callback){
+
+        // https://github.com/substack/node-browserify/issues/1044#issuecomment-72384131
+        var b = browserify(options || {}) // pass options
```
Comment (tnathanjames) [inline styleguide/tools/gulp/pipelines/js.js]: Could be let.

**Review by tnathanjames — APPROVED** [review]
Some of the ES6 updates could be made, but it is probably okay.

**Comment by rbayliss** [comment]
Thanks for the review everyone!  Updated and merged!

**Pull request merged**